### ISSUE TITLE
Incorrect clang-format pointer alignment when variables are declared in conditionals

### ISSUE THREAD

**Issue by JohnC32** [llvm/llvm-project#60146]
Given:

```c++
void foo1(const T1* p1) {
    if (T2* v; check(&v, p1)) {
        goo(v);
    }
}
```

and `_clang-format`:

```yaml
---
BasedOnStyle: Google
IndentWidth: 4
# Google style default is PointerAlignment: Left,
PointerAlignment: Left
...
```

clang-format 15 yields the following. Notice the pointer alignment is no longer left, despite
asking for it to be left.

```c++
void foo1(const T1* p1) {
    if (T2 * v; check(&v, p1)) {
        goo(v);
    }
}
```

In this case, we scoped the variable declaration in the if conditional. If we move it out of the
conditional, pointer alignment is as expected:

```c++
void foo2(const T1* p1) {
    T2* v;
    if (check(&v, p1)) {
        goo(v);
    }
}
```

**Comment by jeremyong** [comment]
I'm seeing this also with an even simpler repro:

```c++
for (int i = 0; Foo& foo : foos)
```

turns into

```c++
for (int i = 0; Foo & foo : foos)
```

even with `PointerAlignment: Left`. I am assuming that in the presence of scoped declaration statements, subsequent `&` operators are parsed as bitwise-and operators.

**Comment by swalkner** [comment]
I can confirm this behaviour and would like to bump this report.

```cpp
if (MyClass * obj{GetMyClass()})
```
But - for some reason - `auto` and primitive-types seem to work as expected:
```cpp
if (auto* obj{GetMyClass()})
```
```cpp
if (int* obj{GetInt()})
```

I have reproduced that in clang-format:
* 17.0.3 (packaged with VS)
* 18.1.3 (and other versions via clang-format-configurator.site)

The only config option I changed:
> PointerAlignment: Left

**Comment by owenca** [comment]
We have the `TypeNames` option for user-defined types when clang-format can't tell if an identifier before */&/&& is a type or variable:
```
$ clang-format -version
clang-format version 19.1.0
$ cat foo.cpp
if (T2 *v; check(&v, p1))
  ;
for (int i = 0; Foo &foo : foos)
  ;
if (MyClass *obj{GetMyClass()})
  ;
$ cat .clang-format
TypeNames: [T2, Foo, MyClass]
$ clang-format foo.cpp | diff foo.cpp -
$ 
```

**Comment by owenca** [comment]
> Technically we dont have any better option besides providing this additional TypeNames?

In general, `TypeNames` can be used if there are no better alternatives. (See https://reviews.llvm.org/D155273 for some background.) Nevertheless, I'll see if something similar to #109229 can be done here.

**Comment by owenca** [comment]
> I'm seeing this also with an even simpler repro:
> 
> ```c++
> for (int i = 0; Foo& foo : foos)
> ```
> 
> turns into
> 
> ```c++
> for (int i = 0; Foo & foo : foos)
> ```
> 
> even with `PointerAlignment: Left`. I am assuming that in the presence of scoped declaration statements, subsequent `&` operators are parsed as bitwise-and operators.

This is related but different. I have moved it to #109358.

**Comment by owenca** [comment]
> I can confirm this behaviour and would like to gently bump this report.
> 
> ```c++
> if (MyClass * obj{GetMyClass()})
> ```
> 
> But - for some reason - `auto` and primitive-types seem to work as expected:
> 
> ```c++
> if (auto* obj{GetMyClass()})
> ```
> 
> ```c++
> if (int* obj{GetInt()})
> ```
> 
> I have reproduced that in clang-format:
> 
> * 17.0.3 (packaged with VS)
> * 18.1.3 (and other versions via clang-format-configurator.site)
> 
> The only config option I changed:
> 
> > PointerAlignment: Left

This is different as there is no semicolon. Please split it to a new issue or use the `TypeNames` option.

**Comment by swalkner** [comment]
Thanks so much for your reply. I created #109371
